plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftSplitManager.java

```diff
@@ -0,0 +1,202 @@
+        }
+        return new RedshiftUnloadSplitSource(executor, connection, statement, unloadOutputLocation, fileSystemFactory.create(session));
+    }
+
+    private String getRedshiftSelectSql(ConnectorSession session, Connection connection, JdbcTableHandle table, List<JdbcColumnHandle> columns)
+            throws SQLException
+    {
+        PreparedQuery preparedQuery = jdbcClient.prepareQuery(session, table, Optional.empty(), columns, ImmutableMap.of());
+        PreparedStatement openTelemetryPreparedStatement = queryBuilder.prepareStatement(jdbcClient, session, connection, preparedQuery, Optional.of(columns.size()));
+        RedshiftPreparedStatement redshiftPreparedStatement = openTelemetryPreparedStatement.unwrap(RedshiftPreparedStatement.class);
+        String selectQuerySql = redshiftPreparedStatement.toString();
+        return queryModifier.apply(session, selectQuerySql);
+    }
+
+    private PreparedStatement buildUnloadSql(ConnectorSession session, Connection connection, List<JdbcColumnHandle> columns, String redshiftSelectSql, String unloadOutputLocation)
+            throws SQLException
+    {
+        String unloadSql = "UNLOAD ('%s') TO '%s' IAM_ROLE '%s' FORMAT PARQUET MAXFILESIZE 64MB MANIFEST VERBOSE".formatted(
+                formatStringLiteral(redshiftSelectSql),
+                unloadOutputLocation,
+                unloadAuthorization.orElseThrow());
+        return queryBuilder.prepareStatement(jdbcClient, session, connection, new PreparedQuery(unloadSql, List.of()), Optional.of(columns.size()));
+    }
+
+    private static String formatStringLiteral(String x)
+    {
+        return x.replace("'", "''");
+    }
+
+    private static boolean isUnloadSupported(JdbcTableHandle table, List<JdbcColumnHandle> columns)
+    {
+        // Nothing to unload as there are no columns to be fetched from Redshift
+        if (table.getColumns().isPresent() && table.getColumns().get().isEmpty()) {
+            return false;
+        }
+        if (containsUnsupportedType(columns)) {
+            return false;
+        }
+        // Unload command doesn't support limit clause
+        if (table.getLimit().isPresent()) {
+            return false;
+        }
+        if (containsFilterConditionOnDecimalTypeColumn(table)) {
+            return false;
+        }
+        return true;
+    }
+
+    // Unsupported unload command data types when using Parquet output file format
+    private static boolean containsUnsupportedType(List<JdbcColumnHandle> columns)
+    {
+        // ERROR: UNLOAD varbyte column "col_0" is only supported for TEXT/CSV.
+        // ERROR: UNLOAD time without time zone column "value" is only supported for TEXT/CSV.
+        return columns.stream().anyMatch(column -> column.getColumnType() instanceof TimeType || column.getColumnType() instanceof VarbinaryType);
+    }
+
+    // Redshift driver generates incorrect cast precision in select query for decimal columns
```

Comment:
Possibly this could be by design from the redshift driver as there isn't any issue if running decimal column query directly through Redshift driver but only `RedshiftPreparedStatement.toString()` generates incorrect cast query. I'd still try opening an issue with redshift drivers.

Reply:
It's a bug even if just the `toString` is wrong. You could try raising PR there if the fix is obvious, but at least report the issue.

Reply:
Created https://github.com/aws/amazon-redshift-jdbc-driver/issues/129